tools/yulPhaser/SimulationRNG.cpp

```diff
@@ -35,16 +35,16 @@ bool SimulationRNG::bernoulliTrial(double _successProbability)
 	return static_cast<bool>(distribution(s_generator));
 }
 
-uint32_t SimulationRNG::uniformInt(uint32_t _min, uint32_t _max)
+size_t SimulationRNG::uniformInt(size_t _min, size_t _max)
 {
-	boost::random::uniform_int_distribution<> distribution(_min, _max);
```

Comment:
[`std::uniform_int_distribution`](https://en.cppreference.com/w/cpp/numeric/random/uniform_int_distribution)? It seems to work with unsigned types too.

Reply:
@a3d4 Originally I chose boost because I wanted to use `random_device` for seeding and `std::random_device` is not portable (and is known to be implemented in a deterministic way in some environments). I later realized that I don't really need it for my use case (i.e. just a simulation) so I could probably switch to `std::random`. I think I have encountered some other issue that made me stick to `boost` though but I don't really remember now. Do you think switching to `std::random` would be worthwhile?

Reply:
Ah, I did not read what @a3d4 wrote :)
@cameel, it is great if you can swap it with `std::`

Reply:
OK. I'll try to swap it then.

Reply:
I just realized what my issue with `std::random` was. Basically this: [How can I get an implementation-agnostic version of std::uniform_int_distribution?](https://stackoverflow.com/questions/44520973/how-can-i-get-an-implementation-agnostic-version-of-stduniform-int-distributio)

So using `--seed` option in `yul-phaser` will no longer guarantee reproducible results across different STL implementations. And my earlier experiment results may not be exactly reproducible either (though this might not be all that important since you can easily break reproducibility in other ways, by changing even minor implementation details of the algorithm).

Actually Boost does not have any strong guarantees for this but there's just one implementation of Boost so in practice the results will be reproducible (though they might not be across Boost versions). That's not the case with STL.

What do you think? Should we switch anyway? Reproducibility is not a hard requirement for phaser to work but it's nice to have. My decision back then was simply "if we depend on Boost anyway, why not use `boost::random`"?

Reply:
@a3d4 if you are fine, I would say(after reading all of this) that we stay with boost::random.

Reply:
@mijovic 
Sure, if we have good reasons, boost is fine.

> Distribution (both uniform and binomial) doesn't work with unsigned types as there is an error when resolving abs(unsigned)

This still puzzles me. How can I reproduce? Just using `boost::random::uniform_int_distribution<size_t> distribution(_min, _max);` compiles, at least.

Reply:
I added a comment in the code explaining why boost is used.

@a3d4
Looks like `uniform_int_distribution<size_t>` works just fine so I'm using that now.

Only binomial has problems. If you change the line to:
```
boost::random::binomial_distribution<size_t> distribution(static_cast<long>(_numTrials), _successProbability);
```
if fails with:
```
/projects/solidity/tools/yulPhaser/SimulationRNG.cpp:55:53:   required from here
/usr/include/boost/random/binomial_distribution.hpp:324:30: error: call of overloaded ‘abs(long unsigned int)’ is ambiguous
```
Apparently the uniform distribution does not use `abs()` and this one does.

Reply:
OK. BTW, it seems that boost 1.73 is happy with `boost::random::binomial_distribution<size_t>`.

Reply:
Ah, no, fails with the same error.

Reply:
Side remark: `boost::random::binomial_distribution<uint16_t>` compiles :).

Reply:
Indeed. And `uint32_t` does not. Probably there is only one possible conversion for a 16-bit int so it's not ambiguous.

Reply:
So, decision would be to stay with `boost::random::binomial_distribution<long>`?

Reply:
I'm for it. `boost` in this case offers reproducible results. And as for `long` vs `size_t`- phaser never uses the full range anyway and if someone tried, it would be caught by the assertion above.

Reply:
> Indeed. And uint32_t does not. Probably there is only one possible conversion for a 16-bit int so it's not ambiguous.

I think the issue is quite simple. Forget boost. Consider
```
#include <cmath>
unsigned x = 0;
abs(a);
```
We get the same error, as there is no `abs(unsigned)` (makes some sense), hence few possible overloads.

With `unsigned short` instead of `unsigned` the argument gets promoted to `int` first. (Which is even not guaranteed, `short unsigned` can have the same size as `int` of course.)